### DISCUSSION TITLE
Why version downgrade is not allowed?

### DISCUSSION THREAD

**Discussion by mogliang** [kubernetes-sigs/cluster-api#10942] [General]
Hi all,

I'm wondering why version downgrade is not allowed for cluster.spec.topology.version? I heard that kubeadm don't support downgrade, but I didn't find any official statement to support that, besides, based on my test, kubeadm do let me to downgrade version.

I know downgrade minor version may fail because depencency like etcd may not support downgrade for certain version (3.2->3.1), but i'm wondering if patch version downgrade is always safe to do? 

Thanks,
Qi

**Comment by neolit123** [comment]
> kubeadm do let me to downgrade version.

use at your own risk. once a new kubeadm version starts managing the cluster if you downgrade to the old version it might no longer understand this cluster and it's objects. this is actually happening in 1.31 where a new v1beta4.ClusterConfiguration will exist in the kubes-syte/kubeadm-config config-map. i think other k8s components are more tolerant, but there are no plans to add downgrade support for kubeadm.  

k8s as a whole does not support downgrade well, so if you google product info for GKE, EKS and other you will see none support downgrade.

> I know downgrade minor version may fail because depencency like etcd may not support downgrade for certain version (3.2->3.1), but i'm wondering if patch version downgrade is always safe to do?

it might work, as long as PATCH bug fix did not introduce a behavior change that is not downgradable.

if you want a more official statement for k8s as a whole, contact:
https://github.com/kubernetes/community/blob/master/sig-architecture/README.md
